### PR TITLE
unordered_map_concurrent performance improvement

### PR DESCRIPTION
**Pull request by fpsunflower**
Use the upper bits of the hash for bin selection instead of hashing the hash.

This is a small but measurable improvement on the benchmark I am looking at now.

### CODE REVIEW

File: src/include/OpenImageIO/unordered_map_concurrent.h [hunk 1]
```diff
@@ -74,13 +74,16 @@ OIIO_NAMESPACE_BEGIN
 ///
 
 template<class KEY, class VALUE, class HASH=std::hash<KEY>,
-         class PRED=std::equal_to<KEY>, size_t BINS=16,
+         class PRED=std::equal_to<KEY>, int LOG2_BINS=4,
```
Comment (lgritz) [inline src/include/OpenImageIO/unordered_map_concurrent.h]: I think this is a confusing change. Can we keep it BINS, use static_assert to ensure it's a power of 2 (if we must), and figure out the bin shift via constexpr and bit twiddling tricks?

Comment (fpsunflower) [reply src/include/OpenImageIO/unordered_map_concurrent.h]: I can give it a try ...

**Comment by fpsunflower** [comment]
Pushed update. I agree its cleaner this way.

**Comment by fpsunflower** [comment]
Tweaked a bit more to protect against potential mistakes in the future.

**Comment by lgritz** [comment]
LGTM

**Comment by lgritz** [comment]
Be sure the CI passes before you merge. Don't want this to fail for some platforms.

**Pull request merged**